src/utils/globalThis.ts

```diff
@@ -1,13 +1,32 @@
-// Implementation of globalThis derived from https://github.com/tc39/proposal-global/blob/master/polyfill.js
+// See https://github.com/paulmillr/es6-shim/commit/2367e0953edd01ae9a5628e1f47cf14b0377a7d6
+// and https://github.com/tc39/proposal-global/blob/269dd7c646451e72e3e3edc573b57ce766aac86a/README.md#rationale
 
-/** @internal */
 // tslint:disable
-const internalGlobalThis: any = (function (global) {
-  return global.globalThis ? global.globalThis : global;
-})(typeof this === 'object' ? this : Function('return this')());
+declare const self: any;
+declare const window: any;
+declare const global: any;
 
 /**
  * Internal polyfill for `globalThis`
  * @internal
  */
-export const getGlobal = () => internalGlobalThis;
+export let getGlobal = () => {
+  const globalThis = (function () {
```

Comment:
I'd rather keep this globalThis outside of the function otherwise its computation will be done over and over.

Here is an illustration in codesandbox.
See: https://codesandbox.io/s/flamboyant-field-zopcx?file=/src/globalThis.spec.ts

Explanation of the codesandbox:
- Go in the test tab.
- I tweaked the code to add a console log just to confirm that it would be executed twice while I expect the result to be cached.

Reply:
The tests work for me.

The code should be executed only once because `getGlobal` replaces itself with `() => globalThis` after running the first time. Think of it as a lazily-evaluated `globalThis` where the detection code runs only when the `getGlobal` function is actually used. If it's never called, the detection code (which potentially throws an Error) never runs.

Reply:
Actually, the replacement will not work in the case of an import. What you'll import is the lazy function, it'll never be updated.

Reply:
Alternatively, we can just don't do the optim. After all, keeping the global is pretty useless given the length and complexity of code.

Reply:
According to ES Module semantics the let binding should update: https://stackoverflow.com/a/32558929

Reply:
Yes, I know. But only if not exported. Let's take this code:

```js
import {getGlobal} from './'
const initialGetGlobal = getGlobal;
getGlobal();
assert.equal(getGlobal, initialGetGlobal); // will not fail, getGlobal will not be changed even with the let trick
```

I'll come back with a repro outside of codesandbox

Reply:
Non-compliant ES module bundler 🤷 . I committed your suggestion.

Reply:
It seems that you are right. 

The problem will not be with esm but cjs (I still support it). An user that does `var { getGlobal } = require("./global.js")` will relaunch the computation on each call without using the cached one.

Just for the record here is the repro for CJS:

```js
// global.js
"use strict";

var _getGlobal = function getGlobal() {
  var globalThis = (function () {
    console.log("iife");
    if (typeof self !== "undefined") return self;
    if (typeof window !== "undefined") return window;
    if (typeof global !== "undefined") return global;
    throw new Error("unable to locate global object");
  })();

  exports.getGlobal = _getGlobal = function getGlobal() {
    return globalThis;
  };

  return globalThis;
};

exports.getGlobal = _getGlobal;

// play.js
"use strict";
const { getGlobal } = require("./global.js");

const initialGetGlobal = getGlobal;
getGlobal();
console.log(initialGetGlobal === getGlobal);
getGlobal();

// output of: 
// $ node play.js 
iife
true
iife
```

No problem with ESM, as you said:

```js
// global.mjs
export let getGlobal = () => {
  const globalThis = (function () {
    console.log("iife");
    if (typeof self !== "undefined") return self;
    if (typeof window !== "undefined") return window;
    if (typeof global !== "undefined") return global;
    throw new Error("unable to locate global object");
  })();
  getGlobal = () => globalThis;
  return globalThis;
};

// play.mjs
import { getGlobal } from "./global.mjs";

const initialGetGlobal = getGlobal;
getGlobal();
console.log(initialGetGlobal === getGlobal);
getGlobal();

// output of:
// $ node play.mjs
iife
false
```

Anyway, I prefer the version with an extracted code, it looks easier.